hexrd/ui/calibration/hedm/calibration_runner.py

```diff
@@ -0,0 +1,564 @@
+
+    def pre_validate(self):
+        # Validation to perform before we do anything else
+        if not self.active_overlays:
+            msg = 'There must be at least one visible rotation series overlay'
+            raise Exception(msg)
+
+        ome_periods = []
+        for overlay in self.active_overlays:
+            options = overlay['options']
+            if any(options.get(x) is None for x in ('eta_width', 'tth_width')):
+                msg = (
+                    'All visible rotation series overlays must have widths '
+                    'enabled'
+                )
+                raise Exception(msg)
+
+            ome_periods.append(options['ome_period'])
+
+        for i in range(1, len(ome_periods)):
+            if not np.allclose(ome_periods[0], ome_periods[i]):
+                msg = (
+                    'All visible rotation series overlays must have '
+                    'identical omega periods'
+                )
+                raise Exception(msg)
+
+        materials = [overlay['material'] for overlay in self.active_overlays]
+        if not all(x == materials[0] for x in materials):
+            msg = (
+                'All visible rotation series overlays must have the same '
+                'material'
+            )
+            raise Exception(msg)
+
+        if not np.any(self.all_flags):
+            msg = 'There are no refinable parameters'
+            raise Exception(msg)
+
+        # Ensure we have omega metadata
+        try:
+            create_indexing_config()
+        except OmegasNotFoundError:
+            msg = (
+                'No omega metadata found. Be sure to import the image '
+                'series using the "Simple Image Series" import tool.'
+            )
+            raise Exception(msg)
+
+
+def parse_spots_data(spots_data, instr, grain_ids, refit_idx=None):
+    hkls = {}
+    xyo_det = {}
+    idx_0 = {}
+    for det_key, panel in instr.detectors.items():
+        hkls[det_key] = []
+        xyo_det[det_key] = []
+        idx_0[det_key] = []
+
+        for ig, grain_id in enumerate(grain_ids):
+            data = spots_data[grain_id][1][det_key]
+            # Convert to numpy array to make operations easier
+            data = np.array(data, dtype=object)
+            valid_reflections = data[:, 0] >= 0
+            not_saturated = data[:, 4] < panel.saturation_level
+
+            if refit_idx is None:
+                idx = np.logical_and(valid_reflections, not_saturated)
+                idx_0[det_key].append(idx)
+            else:
+                idx = refit_idx[det_key][ig]
+                idx_0[det_key].append(idx)
+
+            hkls[det_key].append(np.vstack(data[idx, 2]))
+            meas_omes = np.vstack(data[idx, 6])[:, 2].reshape(sum(idx), 1)
+            xyo_det[det_key].append(np.hstack([np.vstack(data[idx, 7]),
+                                               meas_omes]))
+
+    return hkls, xyo_det, idx_0
```

Comment:
@joelvbernier The function above is a version of [`parse_reflections_tables`](https://github.com/HEXRD/hexrd/blob/5bb755775eea1a076f3018f48923ef74ec6566f7/hexrd/fitting/calibration.py#L351-L402) that runs on `pull_spots()` output rather than on spots files. Can you help me make sure it looks good?